[FLINK-2828] [table] Add interfaces for Table API input formats

twalthr opened the pull request:
This PR implements TableSources as described in FLINK-2828. It is an uncoupling of #1127 for FLINK-2167. All feedback I got from @aljoscha is implemented.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaBatchTranslator.scala:
```diff
@@ -39,7 +40,7 @@ import org.apache.flink.api.table.{ExpressionException, Row, Table}
  * translating them back to Java [[org.apache.flink.api.java.DataSet]]s.
  */
-class JavaBatchTranslator extends PlanTranslator {
+class JavaBatchTranslator(env: ExecutionEnvironment = null) extends PlanTranslator {
```
We could use an `Option` here to avoid `null`, since `null` is evil.

twalthr replied:
You are right. Sorry, Scala beginner's mistake. I will correct that.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaStreamingTranslator.scala:
```diff
@@ -38,7 +40,7 @@ import org.apache.flink.streaming.api.operators.StreamMap
  */
 
-class JavaStreamingTranslator extends PlanTranslator {
+class JavaStreamingTranslator(env: StreamExecutionEnvironment = null) extends PlanTranslator {
```
same here

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/FieldBacktracker.scala:
```diff
@@ -0,0 +1,75 @@
+   * @return original field name with corresponding AdaptiveTableSource or null
+   */
+  def resolveFieldNameAndTableSource(op: PlanNode, fieldName: String):
+      (AdaptiveTableSource, String) = {
```
Can we avoid `null` return values by using `Option[(AdaptiveTableSource, String)]`?

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/FieldBacktracker.scala:
```diff
@@ -0,0 +1,75 @@
+      case s@Select(input, selection) =>
+        var resolvedField: (AdaptiveTableSource, String) = null
+        // only follow unmodified fields
+        selection.foreach {
```
`selection` is of type `Seq[Expression]`, right? But `resolveField` stores only the last resolved field. Is this correct? Shouldn't we make sure that the selection contains only a single element?

twalthr replied:
No, `selection` can have multiple elements. The lines above just make sure that at least one Expression links to the `TableSource`'s field.

tillrohrmann replied:
But with this implementation you will take the resolved value of the last expression in `selection`. All other resolved fields will be discarded. Why don't you simply take the first element of `selection` via `head` and resolve it? Or do you perform any side effects in the resolve methods?

twalthr replied:
You are right the other resolved fields (even valid ones) were discarded and possibly overwritten by `None`, this was a bug. I changed the lines. Now the first resolved field found will be taken.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/FieldBacktracker.scala:
```diff
@@ -0,0 +1,75 @@
+      (AdaptiveTableSource, String) = {
+    op match {
+      case s@Select(input, selection) =>
+        var resolvedField: (AdaptiveTableSource, String) = null
```
Why not returning the resolved field directly from the `selection`?

twalthr replied:
See answer above.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/PredicatePruner.scala:
```diff
@@ -0,0 +1,88 @@
+   * @param expr expression to be copied and pruned
+   * @return new pruned expression
+   */
+  def pruneExpr(filter: (Expression) => Boolean, expr: Expression): Expression = {
```
Parentheses needed for `(Expression)`?

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/FieldBacktracker.scala:
```diff
@@ -0,0 +1,79 @@
+              case _ => // do nothing
+            }
+          }
+        })
```
But here again you will traverse the whole `selection` sequence. Is there any special reason for it? Why not simply calculate the elements which are needed to find the first resolved field? With Scala you have all the possibilities at hand and furthermore you get rid of the `var` variable:

```
val resolvedField = selection.view.flatMap{
  case ResolvedFieldReference(name, _) if name == fieldName =>
    resolvedField = resolveFieldNameAndTableSource(input, fieldName)
  case n@Naming(ResolvedFieldReference(oldName, _), newName) if newName == fieldName =>
    resolvedField = resolveFieldNameAndTableSource(input, oldName)
  case _ => // do nothing
}.headOption
```

twalthr replied:
Sorry, I wasn't aware of the non-strict/strict concepts in Scala so far. Your solution is definitely nicer. I will correct that ;)

twalthr replied:
Build was successful (at least my parts). Do you think I can merge it?

aljoscha commented:
I want to have a final look today. Is the Scala ok from your side now, @tillrohrmann?

tillrohrmann commented:
Yeah looks good now :-)

twalthr commented:
@aljoscha Do you think you'll find time this week for the PR? All Table API input formats are waiting for it...

aljoscha commented:
Yes, sorry for the delay.

aljoscha commented on flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala:
```diff
@@ -64,15 +62,19 @@ import scala.reflect.ClassTag
   /**
    * Gets the config object.
    */
+  @Override
```
In Scala override is done as:
```
override def getConfig ...
```

this is also in StreamExecutionEnvironment.scala

aljoscha commented:
Why is it `AdaptiveTableSource` and `StaticTableSource`? Aren't static sources a subset of adaptive sources that just don't adapt?

I don't think we need to have to interfaces and two ways of handling them in the translators. Or maybe there is another reason that I'm not seeing? :sweat_smile:

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaBatchTranslator.scala:
```diff
@@ -54,6 +55,22 @@ class JavaBatchTranslator extends PlanTranslator {
+      case adaptive: AdaptiveTableSource => Table(Root(adaptive, adaptive.getOutputFields()))
+
+      case static: StaticTableSource =>
+        createTable(static.createStaticDataSet(env.get),
```
I think it's more scalaesque to use Scala's pattern matching to extract the `ExecutionEnvironment` from `env` instead of the `if` check with a following `get`.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaBatchTranslator.scala:
```diff
@@ -118,6 +135,12 @@ class JavaBatchTranslator extends PlanTranslator {
+        if (env.isEmpty) {
+          throw new ExpressionException("This operation requires a TableEnvironment.")
+        }
+        tableSource.createAdaptiveDataSet(env.get)
```
Same here.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaStreamingTranslator.scala:
```diff
@@ -53,6 +55,22 @@ class JavaStreamingTranslator extends PlanTranslator {
+      case adaptive: AdaptiveTableSource => Table(Root(adaptive, adaptive.getOutputFields()))
+
+      case static: StaticTableSource =>
+        createTable(static.createStaticDataStream(env.get),
```
here as well

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/JavaStreamingTranslator.scala:
```diff
@@ -113,8 +131,14 @@ class JavaStreamingTranslator extends PlanTranslator {
+        if (env.isEmpty) {
+          throw new ExpressionException("This operation requires a TableEnvironment.")
+        }
+        tableSource.createAdaptiveDataStream(env.get)
```
`env` extraction

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/java/table/TableEnvironment.scala:
```diff
@@ -45,7 +60,12 @@ class TableEnvironment {
+      case batchTranslator: JavaBatchTranslator =>
+        batchTranslator.createTable(set, fields)
+      case _ =>
+        throw new IllegalArgumentException("ExecutionEnvironment does not support Java DataSets.")
```
Shouldn't it be `DataStreams` instead of `DataSets`?

tillrohrmann replied:
Sorry, forget my comment.

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/PredicatePruner.scala:
```diff
@@ -0,0 +1,88 @@
+        }
+        else {
+          And(prunedLeft, prunedRight)
+        }
```
What if left and right are `NopExpressions`?

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/PredicatePushdown.scala:
```diff
@@ -0,0 +1,120 @@
+      val result = tsExpr.transformPost {
+        case rfr@ResolvedFieldReference(fieldName, typeInfo) =>
+          ResolvedFieldReference(
+            resolveFieldNameAndTableSource(inputOperation, fieldName).get._2,
```
Are we sure that this `Option` is always `Some`?

tillrohrmann commented on flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/expressions/analysis/PredicatePushdown.scala:
```diff
@@ -0,0 +1,120 @@
+      case bc: BinaryComparison if bc.right.isInstanceOf[ResolvedFieldReference] =>
+        val fieldRef = bc.right.asInstanceOf[ResolvedFieldReference]
+        val resolvedField = resolveFieldNameAndTableSource(inputOperation, fieldRef.name)
+        resolvedField.isDefined && resolvedField.get._1 == ts
```
pattern matching is more scalaesque

twalthr commented:
@aljoscha I thought it makes sense to distinguish between `AdaptiveTableSource` and `StaticTableSource`. Calculating the predicates to be pushed down is quite expensive, it should only happen if the TableSource needs it. `AdaptiveTableSource` need to return `Rows` where fields can be null. `StaticTableSource`s can return POJOs or Tuples.
@tillrohrmann thanks again for your feedback, I will correct it.

aljoscha commented:
I think the performance of the pre-flight phase is mostly irrelevant, even if it is slow no-one will notice since it only runs once before the job starts. IMHO, simplicity is key here, and the two different sources require two code-paths to handle them even though the one source provides a subset of the functionality of the other source.

twalthr commented:
Most input formats won't use the adaptive sources. Sure performance is not that important in the pre-flight phase, but if we always think so, the pre-flight phase will some day take minutes. Simplicity is key but IMHO it is simpler from an extension point of view to implement an interface that fits to the need. One type is evaluated immediately the other is lazy evaluated during translation. I think this is differentiation/argument enough for 2 interfaces.

twalthr commented:
@aljoscha we should finally reach a consesus for this PR

aljoscha commented:
Yes, we should, this has been lying around for a while now.

I have been thinking about this for a while now and I'm afraid I cannot recommend to merge this in the current form. This PR is tackling two issues: 1) adding new interfaces for Table API input formats, 2) pushing filter predicates into sources. 1) will end up being a question of style, i.e. whether we want to have two input interfaces, or just one that can do both. 

On 2), I think we need a different approach. The approach taken in this PR is to enhance the Rules that analyze/transform expression with knowledge about the input operation. This allows to backtrack fields to the sources and notify the source of used fields/predicates. I think, however, that this is not the correct level to implement it. The plan of operations is a Tree itself and can also be analyzed/transformed using the Tree framework of the Table API. I think it would be beneficial to provide analysis steps that can actually push filters towards the sources. (And then fuse adaptive source and filter if possible) This will be helpful even if no adaptive sources are present and it will transform one valid tree of the user query into another valid tree of the user query which makes it easy to see the optimizations that are taking place and to debug it.

To show what I mean I implemented a small prototype that prunes unused fields of selection expressions and pushes filters towards sources: https://github.com/aljoscha/flink/tree/table-filter-push. There is an example in there: `FilterPushExample.scala`, the code for tree analysis is in `PlanAnalyzer.scala`. The example query is this:
```
val input = env.fromElements((1, 1, 1), (2, 2, 2), (4, 4, 4), (5, 5, 5))

    input
      .as('a, 'b, 'c)
      .select('a as 'd, 'b, 'c + 15 as 'e)
      .filter('d > 3)
      .select('d)
      .toDataSet[Row].print
```

if the example is run it will print the operation tree after tree analysis phases, this is the printout:
```
Before PushFilter: Select(Filter(Select(Root(MapOperator@49c43f4e,ArraySeq((a,Integer), (b,Integer), (c,Integer))), 'a as 'd,'b,('c + 15) as 'e), 'd > 3), 'd)
After PushFilter: Select(Select(Filter(Root(MapOperator@49c43f4e,ArraySeq((a,Integer), (b,Integer), (c,Integer))), 'a > 3), 'a as 'd,'b,('c + 15) as 'e), 'd)
Before prune: Select(Select(Filter(Root(MapOperator@49c43f4e,ArraySeq((a,Integer), (b,Integer), (c,Integer))), 'a > 3), 'a as 'd,'b,('c + 15) as 'e), 'd)
After prune: Select(Select(Filter(Root(MapOperator@49c43f4e,ArraySeq((a,Integer), (b,Integer), (c,Integer))), 'a > 3), 'a as 'd), 'd)
ROOT used fields: a

```

twalthr commented:
So basically what you want to say is: throw your entire code away.
I'm disappointed that you think about the whole concept now, after I have rebased and reworked the code multiple times. I published a prototype of it in mid-September (#1127), where I wanted to know if the concept is fine. I won't have the resources and motivation to start from scratch.

Yes, you are right. Your solution is the nicer one. However, the question is how much optimizer logic the Table API should have implemented. Actually I wanted to discuss the following under FLINK-2099, but we can also discuss it here. If we base some further developed library such as Calcite on top of the Table API, optimizer logic will be unnecessary in the near future. In a long term perspective it might make sense to convert Table API DSL to Calcite nodes, optimize them, and use the Table API plan nodes only as the physical plan nodes without further optimization. In this case the only thing missing is predicate/projection pushdown from the Table API tree into the input formats. Calcite could do push down for us.

aljoscha commented:
You are right of course and I'm sorry I didn't put in the required time earlier.

Your are also right about the Calcite thing. If we decide to use calcite and have the Table API basically as another fronted to generate SQL plans then the Table API optimization part can be very thin. In that case we would also not need the Predicate push down as implemented as part of this PR, though, and can keep it lean and simple.

aljoscha commented:
@twalthr is this still valid with the changes in the Table API or can it be closed?

twalthr commented:
I will close this issue in favor of the new design `TableDef` design.
